wallet/walletrpc.c

```diff
@@ -1066,3 +1067,113 @@ static const struct json_command sendpsbt_command = {
 };
 
 AUTODATA(json_command, &sendpsbt_command);
+
+static struct command_result *
+json_signmessagewithkey(struct command *cmd, const char *buffer,
+			const jsmntok_t *obj UNNEEDED, const jsmntok_t *params)
+{
+	/* decoding the address */
+	const u8 *scriptpubkey;
+	const char *message;
+
+	/* from wallet BIP32 */
+	struct pubkey pubkey;
+
+	if (!param(
+		cmd, buffer, params,
+                p_req("message", param_string, &message),
+		p_req("address", param_bitcoin_address, &scriptpubkey),
+                NULL))
+		return command_param_failed();
+
+	const size_t script_len = tal_bytelen(scriptpubkey);
+
+	/* FIXME: we already had the address from the input */
+	char *addr;
+	addr = encode_scriptpubkey_to_addr(tmpctx, chainparams, scriptpubkey);
+
+	if (!is_p2wpkh(scriptpubkey, script_len, NULL)) {
+		/* FIXME add support for BIP 322 */
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "Address is not p2wpkh and "
+				    "it is not supported for signing");
+	}
+
+	u32 keyidx = UINT32_MAX;
+
+	/* loop over all generated keys, find a matching key */
+	/* FIXME: alternatively, can we use the our_addresses hash table?
+	 *struct script_with_len *key;
+	 *struct wallet_address *addr =
+	 *    wallet_address_htable_get(cmd->ld->wallet->our_addresses, key);
+	 */
+	struct issued_address_type *listaddrtypes =
+	    wallet_list_addresses(tmpctx, cmd->ld->wallet, 1, NULL);
```

Comment:
the db doesn't contain types for old keys.  You need to iterate from 1 to `db_get_intvar(cmd->ld->wallet->db, "bip32_max_index", 0);` to get all keys.

You *could* use `wallet_get_addrtype()` once you've found a match, and refuse if we never issued that keyidx as a p2wpkh?

Reply:
Thanks.
Why is that? Is it a matter of backwards compatibility? I imagine before taproot addresses you
had only p2wpkh types and there was no reason why to have an `addrtype` field.
So that nodes that where running before `addrtype` was introduced have older db entries without the `addrtype`
field?

Anyways, listing all bip32 derived keys until "bip32_max_index" is in fact more robust.

Reply:
[Fixed](https://github.com/ElementsProject/lightning/compare/cbc9205bdc39a0285fd055fa3082466c31941b54..0fcdb2ac5796bd62bd0d0b17f4f7b8696a27a9aa)